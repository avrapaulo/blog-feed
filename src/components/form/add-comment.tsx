import { useRef, useState } from 'react'
import { Form, Button } from 'react-bootstrap'
import { trigger } from 'swr'
import Cookies from 'js-cookie'

interface Props {
  postId: number
}

interface Errors {
  isInvalid: boolean
  message: string
}

const AddComment = ({ postId }: Props) => {
  const inputEl = useRef(null)
  const [form, setForm] = useState<Errors>({ isInvalid: false, message: '' })
  const handleClick = async (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault()
    const content = inputEl.current.value
    const userName = Cookies.get('name')

    if (content === '') {
      setForm({ isInvalid: true, message: 'required field' })
      return
    }

    if (content.length < 6) {
      setForm({ isInvalid: true, message: 'Needs to be at least 6 characters' })
      return
    }

    if (!userName) {
      setForm({ isInvalid: true, message: 'You need to Sign in' })
      return
    }

    await fetch(`api/posts/${postId}/comments`, {
      body: JSON.stringify({
        content,
        user: userName
      }),
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST'
    })

    trigger(`api/posts/${postId}/comments`)

    inputEl.current.value = ''
  }

  const resetForm = async (e: React.ChangeEvent<HTMLElement>) => {
    e.preventDefault()
    setForm({ isInvalid: false, message: '' })
  }

  return (
    <Form className="mt-5">
      <Form.Group>
        <Form.Label htmlFor="inlineFormInputName2" srOnly>
          Comment
        </Form.Label>
        <Form.Control
          className="mb-2 mr-sm-2"
          id="inlineFormInputName2"
          placeholder="Leave here your opinion"
          ref={inputEl}
          isInvalid={form.isInvalid}
          onChange={resetForm}
        />
        <Form.Control.Feedback type="invalid">{form.message}</Form.Control.Feedback>
      </Form.Group>
      <Button type="submit" className="mb-2" onClick={handleClick}>
        Submit
      </Button>
    </Form>
  )
}

export { AddComment }
