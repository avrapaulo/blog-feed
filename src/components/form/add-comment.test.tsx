import { render, fireEvent } from 'test/test-utils'
import { AddComment } from './add-comment'

describe('Add comment component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<AddComment postId={1} />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('matches snapshot with submit', () => {
    const { asFragment, getByText } = render(<AddComment postId={1} />, {})
    fireEvent.click(getByText('Submit'))
    expect(asFragment()).toMatchSnapshot()
  })
})
