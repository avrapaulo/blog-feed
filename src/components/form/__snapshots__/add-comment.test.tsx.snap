// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add comment component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="mt-5"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label sr-only"
        for="inlineFormInputName2"
      >
        Comment
      </label>
      <input
        class="mb-2 mr-sm-2 form-control"
        id="inlineFormInputName2"
        placeholder="Leave here your opinion"
      />
      <div
        class="invalid-feedback"
      />
    </div>
    <button
      class="mb-2 btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`Add comment component matches snapshot with submit 1`] = `
<DocumentFragment>
  <form
    class="mt-5"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label sr-only"
        for="inlineFormInputName2"
      >
        Comment
      </label>
      <input
        class="mb-2 mr-sm-2 form-control is-invalid"
        id="inlineFormInputName2"
        placeholder="Leave here your opinion"
      />
      <div
        class="invalid-feedback"
      >
        required field
      </div>
    </div>
    <button
      class="mb-2 btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
